test/Microsoft.DotNet.Tools.Tests.Utilities/Mock/FileSystemMockBuilder.cs

```diff
@@ -115,170 +420,304 @@ public Stream OpenRead(string path)
-                _files.Remove(source);
-                _files[destination] = content;
+                (DirectoryNode sourceParent, FileNode sourceFileNode)
+                    = _files.GetParentDirectoryAndFileNode(
+                        source,
+                        () => throw new FileNotFoundException($"Could not find file '{source}'"));
+
+                if (_files.TryGetNodeParent(destination, out DirectoryNode current) && current != null)
+                {
+                    current.Subs.Add(new PathModel(destination).FileOrDirectoryName(), sourceFileNode);
+                    sourceParent.Subs.Remove(new PathModel(source).FileOrDirectoryName());
+                }
+                else
+                {
+                    throw new DirectoryNotFoundException($"Could not find a part of the path {destination}");
+                }
             }
 
-            public void Delete(string path)
+            public void Copy(string source, string destination)
             {
-                if (!Exists(path))
+                if (source == null)
                 {
-                    return;
+                    throw new ArgumentNullException(nameof(source));
                 }
 
-                _files.Remove(path);
+                if (destination == null)
+                {
+                    throw new ArgumentNullException(nameof(destination));
+                }
+
+                (_, FileNode sourceFileNode) = _files.GetParentDirectoryAndFileNode(source,
+                    () => throw new UnauthorizedAccessException($"Access to the path {source} is denied")
+                );
+
+                if (_files.TryGetNodeParent(destination, out DirectoryNode current) && current != null)
+                {
+                    if (current.Subs.ContainsKey(new PathModel(destination).FileOrDirectoryName()))
+                    {
+                        throw new IOException($"Path {destination} already exists");
+                    }
+
+                    current.Subs.Add(new PathModel(destination).FileOrDirectoryName(),
+                        new FileNode(sourceFileNode.Content));
+                }
+                else
+                {
+                    throw new DirectoryNotFoundException($"Could not find a part of the path {destination}");
+                }
             }
 
-            public void Copy(string source, string destination)
+            public void Delete(string path)
             {
-                if (!Exists(source))
+                if (_files.TryGetNodeParent(path, out DirectoryNode current))
                 {
-                    throw new FileNotFoundException("source does not exist.");
+                    PathModel pathModel = new PathModel(path);
+                    if (current.Subs.ContainsKey(pathModel.FileOrDirectoryName()))
+                    {
+                        current.Subs.Remove(pathModel.FileOrDirectoryName());
+                    }
```

Comment:
Remove doesn't need to be guarded by ContainsKey, it just returns false when no element is found to remove. This also reduces 2 dictionary lookups to 1.

Reply:
Fixed as suggested